File: Ruby/problem12.rb
require "EulerExtensions"

FACTORS = 500

def triangle_for(n)
  (n * (n+1))/2
end

found = false
curr = 2

while(!found)
  triangle = triangle_for(curr)
  found = triangle.divisors.length > FACTORS
  curr += 1
end

puts triangle

File: Clojure/problem12.clj
(load "euler-common")

(defn triangle [x]
	(loop [curr x sum 0]
		(if (= curr 0)
			sum
			(recur (dec curr) (+ curr sum)))))

(defn divisor-count [x]
	(->>
		x
		factors
		power-set
		distinct
		count))

(defn find-triangle-divisors-greater [x]
	(loop [[t & rest] (map triangle (iterate inc 1))]
		(if (> (divisor-count t) x)
			t
			(recur rest))))

(println (find-triangle-divisors-greater 500))